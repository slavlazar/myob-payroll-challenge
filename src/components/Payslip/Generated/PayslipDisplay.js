import React from 'react';
import PropTypes from 'prop-types';

const payslipKeyMapper = {
  name: 'Employee',
  payPeriod: 'Pay Period',
  payFrequency: 'Pay Frequency',
  annualSalary: 'Annual Income',
  grossIncome: 'Gross Income',
  incomeTax: 'Income Tax',
  netIncome: 'Net Income',
  superTotal: 'Super',
  pay: 'Pay',
};

/**
 * Render prop component that generates key/value pairing of
 * row data for displaying payslip data in both web and PDF views
 * @param {object} Object of data generated from payslip
 * containing keys from payslipKeyMapper and values generated
 * And children which is a function that accepts render props
 * of the calculated column name from key mapper, along with
 * value from data
 */
const PayslipDisplay = ({ data, children }) => (
  <React.Fragment>
    {Object.keys(payslipKeyMapper).map(rowKey => {
      return children({
        key: payslipKeyMapper[rowKey],
        value: data[rowKey],
        _rowKey: rowKey,
      });
    })}
  </React.Fragment>
);

PayslipDisplay.propTypes = {
  /** Children of render props component */
  children: PropTypes.func.isRequired,
  /** Payslip data generated */
  data: PropTypes.object.isRequired,
};

export default PayslipDisplay;
