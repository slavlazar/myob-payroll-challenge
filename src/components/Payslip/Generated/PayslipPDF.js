import React from 'react';
import PropTypes from 'prop-types';
import {
  Page,
  Text,
  View,
  Document,
  StyleSheet,
  Font,
  Image,
} from '@react-pdf/renderer';

import PayslipDisplay from './PayslipDisplay';
import theme from '../../../theme';
import logo from '../../../assets/img/MYOB_logo_RGB_1475x779.png';

Font.register({
  family: 'NeutroMYOB',
  fonts: [
    {
      src: 'https://assets.digital.myob.com/fonts/NeutroMYOB-Regular.woff',
      fontStyle: 'normal',
      fontWeight: 400,
    },
    {
      src: 'https://assets.digital.myob.com/fonts/NeutroMYOB-Medium.woff',
      fontStyle: 'normal',
      fontWeight: 600,
    },
    {
      src: 'https://assets.digital.myob.com/fonts/NeutroMYOB-Demi.woff',
      fontStyle: 'normal',
      fontWeight: 700,
    },
  ],
});

const styles = StyleSheet.create({
  page: {
    paddingTop: 35,
    paddingBottom: 65,
    paddingHorizontal: 35,
    backgroundColor: theme.colour.background,
  },
  logo: {
    width: 120,
    alignSelf: 'flex-end',
  },
  container: {
    backgroundColor: '#ffffff',
    borderTopLeftRadius: 3,
    borderTopRightRadius: 3,
    borderBottomRightRadius: 3,
    borderBottomLeftRadius: 3,
    padding: 20,
  },
  heading: {
    fontFamily: 'NeutroMYOB',
    fontSize: 17,
    fontWeight: 600,
    colour: theme.colour.heading,
    paddingBottom: 16,
  },
  tableContainer: {
    marginVertical: 10,
    flexDirection: 'column',
    flexGrow: 1,
  },
  viewRow: {
    flexDirection: 'row',
    paddingVertical: 10,
    borderBottomColor: '#DCDFE1',
    borderBottomWidth: 1,
    borderBottomStyle: 'solid',
    alignItems: 'center',
    justifyContent: 'flex-start',
  },
  viewRowHeader: {
    borderBottomWidth: 2,
  },
  tableHeader: {
    fontWeight: 600,
  },
  item: {
    alignSelf: 'flex-start',
    width: '40%',
  },
  detail: {
    alignSelf: 'flex-end',
    maxWidth: '55%',
  },
  text: {
    fontFamily: 'NeutroMYOB',
    fontSize: 14,
    fontWeight: 400,
    color: theme.colour.text,
  },
});

/**
 * Generates a PDF version of payslip data
 * @param {object} Payslip data generated
 * @exports pdf
 */
const PayslipPdf = ({ payslip }) => (
  <Document>
    <Page size="A4" style={styles.page}>
      <Image src={logo} style={styles.logo} />
      <View style={styles.container}>
        <Text id="payslip-pdf-name" style={styles.heading}>
          Pay Slip for {payslip.name}
        </Text>
        <View style={styles.tableContainer}>
          <View style={[styles.viewRow, styles.viewRowHeader]}>
            <Text style={[styles.item, styles.text, styles.tableHeader]}>
              Item
            </Text>
            <Text style={[styles.detail, styles.text, styles.tableHeader]}>
              Employee Details
            </Text>
          </View>
          <PayslipDisplay data={payslip}>
            {({ key, value, _rowKey }) => (
              <View
                style={styles.viewRow}
                key={key}
                className="payslip-generated-row"
              >
                <Text
                  style={[styles.text, styles.item]}
                  className="payslip-generated-row--key"
                  data-key={_rowKey}
                >
                  {key}
                </Text>
                <Text
                  style={[styles.text, styles.detail]}
                  className="payslip-generated-row--value"
                >
                  {value}
                </Text>
              </View>
            )}
          </PayslipDisplay>
        </View>
      </View>
    </Page>
  </Document>
);

PayslipPdf.propTypes = {
  /** Payslip object containing all generated data for payslip */
  payslip: PropTypes.shape({
    /** Full name of employee */
    name: PropTypes.string.isRequired,
    /** Pay period of employee */
    payPeriod: PropTypes.string.isRequired,
    /** Annual salary of employee */
    annualSalary: PropTypes.string.isRequired,
    /** Calculated Gross Income of employee */
    grossIncome: PropTypes.string.isRequired,
    /** Calculated Income Tax of employee */
    incomeTax: PropTypes.string.isRequired,
    /** Calculated Net Income of employee */
    netIncome: PropTypes.string.isRequired,
    /** Calculated Super Total of employee */
    superTotal: PropTypes.string.isRequired,
    /** Calculated Pay of employee */
    pay: PropTypes.string.isRequired,
    payFrequency: PropTypes.string,
  }).isRequired,
};

export default PayslipPdf;
