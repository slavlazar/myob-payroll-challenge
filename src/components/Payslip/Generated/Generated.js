import React from 'react';
import PropTypes from 'prop-types';
import { PDFDownloadLink } from '@react-pdf/renderer';

import { Heading, Button, Table, Flex, Card } from '../../shared/ui';
import PayslipPDF from './PayslipPDF';
import PayslipDisplay from './PayslipDisplay';

/**
 * Generated Payslip component displaying all values generated from action
 * @param {object} Payslip object of values and handleReset action
 */
const GeneratedPayslip = ({ payslip, handleReset }) => (
  <Card>
    <Heading id="payslip-generated-name" type="h3">
      Pay Slip for {payslip.name}
    </Heading>
    <Table.Responsive>
      <Table>
        <Table.THead>
          <Table.TR>
            <Table.TH>Item</Table.TH>
            <Table.TH>Employee Details</Table.TH>
          </Table.TR>
        </Table.THead>
        <Table.TBody>
          <PayslipDisplay data={payslip}>
            {({ key, value, _rowKey }) => (
              <Table.TR key={key} className="payslip-generated-row">
                <Table.TD
                  className="payslip-generated-row--key"
                  data-key={_rowKey}
                >
                  {key}
                </Table.TD>
                <Table.TD className="payslip-generated-row--value">
                  {value}
                </Table.TD>
              </Table.TR>
            )}
          </PayslipDisplay>
        </Table.TBody>
      </Table>
    </Table.Responsive>
    <Flex justify="space-between" align="center">
      <Button
        id="payslip-generated-reset-form"
        type="button"
        colour="secondary"
        size="small"
        onClick={handleReset}
        aria-label="Start over and generate new payslip"
      >
        Start over
      </Button>
      <PDFDownloadLink
        id="payslip-generated-pdf-download"
        document={<PayslipPDF payslip={payslip} />}
        fileName={`MYOB_Payslip_${payslip.payPeriod.replace(/\s/g, '_')}.pdf`}
      >
        {({ loading }) =>
          loading ? null : (
            <Button
              size="small"
              type="button"
              colour="primary"
              aria-label="Download generated payslip now"
            >
              Download now
            </Button>
          )
        }
      </PDFDownloadLink>
    </Flex>
  </Card>
);

GeneratedPayslip.propTypes = {
  /** Payslip object containing all generated data for payslip */
  payslip: PropTypes.shape({
    /** Full name of employee */
    name: PropTypes.string.isRequired,
    /** Pay period of employee */
    payPeriod: PropTypes.string.isRequired,
    /** Annual salary of employee */
    annualSalary: PropTypes.string.isRequired,
    /** Calculated Gross Income of employee */
    grossIncome: PropTypes.string.isRequired,
    /** Calculated Income Tax of employee */
    incomeTax: PropTypes.string.isRequired,
    /** Calculated Net Income of employee */
    netIncome: PropTypes.string.isRequired,
    /** Calculated Super Total of employee */
    superTotal: PropTypes.string.isRequired,
    /** Calculated Pay of employee */
    pay: PropTypes.string.isRequired,
    payFrequency: PropTypes.string,
  }).isRequired,
  /** Handle Reset function to go back to Generator screen */
  handleReset: PropTypes.func.isRequired,
};

GeneratedPayslip.defaultProps = {};

export default GeneratedPayslip;
