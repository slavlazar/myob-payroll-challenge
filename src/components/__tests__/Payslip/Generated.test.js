import React from 'react';
import { shallow } from 'enzyme';
import sinon from 'sinon';
import { Font, StyleSheet } from '@react-pdf/renderer';

import Generated from '../../Payslip/Generated/Generated';
import PayslipDisplay from '../../Payslip/Generated/PayslipDisplay';
import PayslipPDF from '../../Payslip/Generated/PayslipPDF';

const defaultProps = {
  payslip: {
    name: 'Bobs Burgers',
    payPeriod: '01 March to 31 March',
    annualSalary: '$60,050.00',
    grossIncome: '$5,004.00',
    incomeTax: '$871.00',
    netIncome: '$4,082.00',
    superTotal: '$450.00',
    pay: '$3,632.00',
    payFrequency: 'Monthly',
  },
  handleReset: sinon.spy(),
};

const shallowGeneratedWrapper = (props = {}) => {
  const allProps = {
    ...defaultProps,
    ...props,
  };

  return shallow(<Generated {...allProps} />);
};

const getDisplayRowValues = wrapped => {
  const display = wrapped.find(PayslipDisplay).shallow();
  const values = display.find('.payslip-generated-row').reduce((accum, row) => {
    const result = accum;
    result[row.find('.payslip-generated-row--key').prop('data-key')] = row
      .find('.payslip-generated-row--value')
      .text();

    return result;
  }, {});

  return values;
};

describe('Components: Payslip: Generated', () => {
  afterEach(() => {
    defaultProps.handleReset.resetHistory();
  });

  describe('PayslipDisplay', () => {
    it('renders correctly', () => {
      const wrapped = shallow(
        <PayslipDisplay data={defaultProps.payslip}>
          {obj => JSON.stringify(obj)}
        </PayslipDisplay>
      );

      expect(wrapped).toMatchSnapshot();
    });
  });

  describe('Generated', () => {
    let wrapped;

    beforeAll(() => {
      wrapped = shallowGeneratedWrapper();
    });

    it('renders correctly', () => {
      expect(wrapped).toMatchSnapshot();
    });

    it('should display the correct title for the user', () => {
      expect(
        wrapped
          .find('#payslip-generated-name')
          .dive()
          .text()
      ).toEqual('Pay Slip for Bobs Burgers');
    });

    it('should handle a reset form action', () => {
      expect(defaultProps.handleReset.callCount).toEqual(0);
      wrapped.find('#payslip-generated-reset-form').simulate('click');
      expect(defaultProps.handleReset.callCount).toEqual(1);
    });

    it('should generate a PDF download link with the correct filename', () => {
      const pfdDownloadLink = wrapped.find('#payslip-generated-pdf-download');
      expect(pfdDownloadLink).toHaveLength(1);
      expect(pfdDownloadLink.prop('fileName')).toEqual(
        'MYOB_Payslip_01_March_to_31_March.pdf'
      );
    });

    it('should display the correct data table information', () => {
      const values = getDisplayRowValues(wrapped);

      expect(values).toMatchObject(defaultProps.payslip);
    });
  });

  describe('Payslip PDF', () => {
    let wrapped;

    beforeAll(() => {
      sinon.stub(Font, 'register');
      sinon.stub(StyleSheet, 'create');

      wrapped = shallow(<PayslipPDF payslip={defaultProps.payslip} />);
    });

    it('renders correctly', () => {
      expect(wrapped).toMatchSnapshot();
    });

    it('should display the correct title for the user', () => {
      expect(wrapped.find('#payslip-pdf-name').text()).toEqual(
        'Pay Slip for Bobs Burgers'
      );
    });

    it('should display the correct data table information', () => {
      const values = getDisplayRowValues(wrapped);

      expect(values).toMatchObject(defaultProps.payslip);
    });
  });
});
