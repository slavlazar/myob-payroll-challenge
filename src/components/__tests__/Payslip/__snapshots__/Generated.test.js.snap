// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Components: Payslip: Generated Generated renders correctly 1`] = `
<Card
  loading={false}
>
  <Heading
    id="payslip-generated-name"
    noMargin={false}
    textAlign="left"
    type="h3"
  >
    Pay Slip for 
    Bobs Burgers
  </Heading>
  <styled.div>
    <styled.table>
      <styled.thead>
        <styled.tr>
          <styled.th>
            Item
          </styled.th>
          <styled.th>
            Employee Details
          </styled.th>
        </styled.tr>
      </styled.thead>
      <styled.tbody>
        <PayslipDisplay
          data={
            Object {
              "annualSalary": "$60,050.00",
              "grossIncome": "$5,004.00",
              "incomeTax": "$871.00",
              "name": "Bobs Burgers",
              "netIncome": "$4,082.00",
              "pay": "$3,632.00",
              "payFrequency": "Monthly",
              "payPeriod": "01 March to 31 March",
              "superTotal": "$450.00",
            }
          }
        >
          <Component />
        </PayslipDisplay>
      </styled.tbody>
    </styled.table>
  </styled.div>
  <styled.div
    align="center"
    justify="space-between"
  >
    <Button
      aria-label="Start over and generate new payslip"
      block={false}
      colour="secondary"
      disabled={false}
      id="payslip-generated-reset-form"
      onClick={[Function]}
      size="small"
      type="button"
    >
      Start over
    </Button>
    <PDFDownloadLink
      document={
        <PayslipPdf
          payslip={
            Object {
              "annualSalary": "$60,050.00",
              "grossIncome": "$5,004.00",
              "incomeTax": "$871.00",
              "name": "Bobs Burgers",
              "netIncome": "$4,082.00",
              "pay": "$3,632.00",
              "payFrequency": "Monthly",
              "payPeriod": "01 March to 31 March",
              "superTotal": "$450.00",
            }
          }
        />
      }
      fileName="MYOB_Payslip_01_March_to_31_March.pdf"
      id="payslip-generated-pdf-download"
    >
      <Component />
    </PDFDownloadLink>
  </styled.div>
</Card>
`;

exports[`Components: Payslip: Generated Payslip PDF renders correctly 1`] = `
<DOCUMENT>
  <PAGE
    size="A4"
    style={
      Object {
        "backgroundColor": "#f3f4f5",
        "paddingBottom": 65,
        "paddingHorizontal": 35,
        "paddingTop": 35,
      }
    }
  >
    <IMAGE
      src="MYOB_logo_RGB_1475x779.png"
      style={
        Object {
          "alignSelf": "flex-end",
          "width": 120,
        }
      }
    />
    <VIEW
      style={
        Object {
          "backgroundColor": "#ffffff",
          "borderBottomLeftRadius": 3,
          "borderBottomRightRadius": 3,
          "borderTopLeftRadius": 3,
          "borderTopRightRadius": 3,
          "padding": 20,
        }
      }
    >
      <TEXT
        id="payslip-pdf-name"
        style={
          Object {
            "colour": "#2e3e4f",
            "fontFamily": "NeutroMYOB",
            "fontSize": 17,
            "fontWeight": 600,
            "paddingBottom": 16,
          }
        }
      >
        Pay Slip for 
        Bobs Burgers
      </TEXT>
      <VIEW
        style={
          Object {
            "flexDirection": "column",
            "flexGrow": 1,
            "marginVertical": 10,
          }
        }
      >
        <VIEW
          style={
            Array [
              Object {
                "alignItems": "center",
                "borderBottomColor": "#DCDFE1",
                "borderBottomStyle": "solid",
                "borderBottomWidth": 1,
                "flexDirection": "row",
                "justifyContent": "flex-start",
                "paddingVertical": 10,
              },
              Object {
                "borderBottomWidth": 2,
              },
            ]
          }
        >
          <TEXT
            style={
              Array [
                Object {
                  "alignSelf": "flex-start",
                  "width": "40%",
                },
                Object {
                  "color": "#515e6d",
                  "fontFamily": "NeutroMYOB",
                  "fontSize": 14,
                  "fontWeight": 400,
                },
                Object {
                  "fontWeight": 600,
                },
              ]
            }
          >
            Item
          </TEXT>
          <TEXT
            style={
              Array [
                Object {
                  "alignSelf": "flex-end",
                  "maxWidth": "55%",
                },
                Object {
                  "color": "#515e6d",
                  "fontFamily": "NeutroMYOB",
                  "fontSize": 14,
                  "fontWeight": 400,
                },
                Object {
                  "fontWeight": 600,
                },
              ]
            }
          >
            Employee Details
          </TEXT>
        </VIEW>
        <PayslipDisplay
          data={
            Object {
              "annualSalary": "$60,050.00",
              "grossIncome": "$5,004.00",
              "incomeTax": "$871.00",
              "name": "Bobs Burgers",
              "netIncome": "$4,082.00",
              "pay": "$3,632.00",
              "payFrequency": "Monthly",
              "payPeriod": "01 March to 31 March",
              "superTotal": "$450.00",
            }
          }
        >
          <Component />
        </PayslipDisplay>
      </VIEW>
    </VIEW>
  </PAGE>
</DOCUMENT>
`;

exports[`Components: Payslip: Generated PayslipDisplay renders correctly 1`] = `
<Fragment>
  {"key":"Employee","value":"Bobs Burgers","_rowKey":"name"}
  {"key":"Pay Period","value":"01 March to 31 March","_rowKey":"payPeriod"}
  {"key":"Pay Frequency","value":"Monthly","_rowKey":"payFrequency"}
  {"key":"Annual Income","value":"$60,050.00","_rowKey":"annualSalary"}
  {"key":"Gross Income","value":"$5,004.00","_rowKey":"grossIncome"}
  {"key":"Income Tax","value":"$871.00","_rowKey":"incomeTax"}
  {"key":"Net Income","value":"$4,082.00","_rowKey":"netIncome"}
  {"key":"Super","value":"$450.00","_rowKey":"superTotal"}
  {"key":"Pay","value":"$3,632.00","_rowKey":"pay"}
</Fragment>
`;
